server/controllers/embed.controller.js

```diff
@@ -8,35 +8,37 @@ import {
 } from '../utils/previewGeneration';
 import { get404Sketch } from '../views/404Page';
 
-export function serveProject(req, res) {
-  Project.findById(req.params.project_id)
-    .exec((err, project) => {
-      if (err || !project) {
-        get404Sketch(html => res.send(html));
-        return;
-      }
-      // TODO this does not parse html
-      const { files } = project;
-      const htmlFile = files.find(file => file.name.match(/\.html$/i)).content;
-      const filesToInject = files.filter(file => file.name.match(/\.(js|css)$/i));
-      injectMediaUrls(filesToInject, files, req.params.project_id);
+export async function serveProject(req, res) {
+  try {
```

Comment:
i'm not a fan of the `try` body being huge here. it's only here to catch errors from mongodb, right? so only line 13 should be in the body.

Reply:
Well, it also catches all errors that are unintended to make sure that the API user gets a response (500) when the server has a problem, instead of just having the connection hang. In non-async handlers this is handled by default by express because they try/catch all the handlers, but express doesn't support `async` natively. There is [express-async-handler](https://www.npmjs.com/package/express-async-handler) which would help us wrap these functions correctly to catch those errors, but you're just moving where the wrapping handler is.

Reply:
wait, i'm confused. are you saying that express doesn't support `async/await` in the handler functions?

Reply:
Yeah, that's right, it doesn't natively handle it, it expects the user to either throw errors synchronously, or pass them all to `next()` manually. Indeed if you look at [the source of that `async` wrapper](https://github.com/Abazhenov/express-async-handler/blob/master/index.js), it is essentially just doing this large `try`/`catch` there. It's a shame that express doesn't understand Promises, but the people behind express decided it'd be too dramatic a change for express to support them. (As an aside they actually decided to redesign express entirely and made [Koa](https://koajs.com/) which is great but differs in some ways.)

So, that leaves you with two options:
1. Don't use `async`/`await` and lose the great syntactical bonuses that gives you, keeping in mind that you still need to handle the asynchronous errors in your promise chains by eventually passing them to `next()`
2. Do use `async`/`await` and be stuck with wrapping each route's handler in some sort of wrapper, be that a library one or just a big `try`/`catch`. The standard `try`/`catch` is something akin to:
```javascript
async function (req, res, next) {
  try {
    // Really long function here
  } catch(e) {
    next(e);
  }
}
```

Reply:
wow, okay. i had no idea! thanks for the detailed explanation. my instinct is to use the express-async-handler middleware, as wrapping everything in a try/catch feels like it gets redundant fast. just to make sure i totally understand this, say the middleware were included in the project, and i wanted to catch errors specific to the mongodb query/connection. that would look like 
```js
let project;
try {
  project = await Project.findById(req.params.project_id);
} catch(e) {
  // do something with the error, e.g. return 404 or 500
}
```

Reply:
Yep, that'd work fine, and you can naturally optionally chose to rethrow the error in certain circumstances for it to be caught by express:
 ```javascript
async function (req, res, next) {
  // Calculate projectParams
  let project;
  try {
    project = await Project.create(projectParams);
  } catch(e) {
    if (!(e instanceof mongoose.Error.ValidationError)) {
      throw e; // rethrow unexpected exceptions so Express will create a 500
    }
    // Forward validation errors to clients
  }
}
```

Reply:
I think that's a very good error handling approach , we can wrap out controllers inside a function which takes out route methods as callbacks and run in try / catch blocks and call next if error occurs , not we have to design custom error handlers to catch different errors (say 404/500)

Reply:
okay great! to me, this feels like the cleanest approach.

Reply:
@catarak I am somewhat confused which message are you referring to , can you give me points on what needs to be done next so I can make incremental prs to achieve that

Reply:
@siddhant1 sure! i think that it would be best to add `express-async-handler` to this project, so that the entire body of the handler functions doesn't have to be a `try` body. otherwise you can keep doing what you were doing.

Reply:
@catarak There are some scenarios where express-async-handler doesn't work as expected.

@siddhant1 check @awaitjs/express before starting work.
Check this blog: http://thecodebarbarian.com/introducing-await-js-express-async-support-for-express-apps

Reply:
@ganeshpatro321 thanks for the update! that's super frustrating. i'm starting to feel like nested callbacks is better than this mess...

Reply:
The only problem with callbacks is not nesting or callbacks hell , I think if not now , we should definitely consider this as an important task in the future.

Reply:
@siddhant1 i agree! i'm not a huge fan of the nested callbacks. however, it feels like the tools to move everything to async/await are in flux, and i think it's best to wait a little bit. 

one option could be to just move to promises. but, honestly, and maybe this is just me, but i'm not sure the promise syntax is too much cleaner/easier to read without async/await

Reply:
I think it's important to note that one of the major benefits of using Promises/(async/await) is the ability to have errors bubble up naturally, rather than not being sure whether some errors may just be being silently ignored instead of raised correctly. As discussed before, Koa is a viable Express replacement that is functionally identical except for supporting promises natively. I would be happy to do an exploration PR into swapping out for that if you think it makes sense to have a look at it @catarak ?

Reply:
that is a great point.

it seems like koa is a fairly radical departure from express, and would involve some major changes. for me, there are higher priority issues, so working on this change would have to be fairly self-directed. however, i am happy to test and give support when i can. if you think there's a possibility of significantly improved error-handling and major code clean up, it's definitely worth investigating!